### PR TITLE
added shutdown capabilities for significant strategy loss

### PR DESCRIPTION
**Pull request by wer32dr**

### CODE REVIEW

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 1]
```diff
@@ -9,36 +9,46 @@ import {SafeInitializable} from "../upgradeable/SafeInitializable.sol";
...
 abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
-    event HealthCheckChanged(address);
-    event HealthCheckEnabledChanged(bool);
+    event HealthCheckChanged(address healthCheck);
+    event ShutdownLossRatioChanged(uint256 ratio);
+    event HealthCheckEnabledChanged(bool enabled);
+    event HealthCheckTriggered(uint8 result);
+
+    // represents 100%
+    uint256 public constant MAX_BPS = 10_000;
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/HealthChecker.sol:18]: Are constants not create memory shifts?

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 2]
```diff
@@ -86,17 +100,76 @@ abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
...
+        } else if(loss > 0 && loss <= totalDebt * shutdownLossRatio / MAX_BPS) { //todo verify the condition
+            // if loss is positive but below critical loss threshold
+            result = 1;
+        } else {
+            // if loss is positive but still makes sense to run transaction to perform reporting.
+            result = 2;
+        }
+
+        emit HealthCheckTriggered(result);
+        return result;
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/HealthChecker.sol:173]: Please make a separate contract and move this logic into this HealthChecker contract. We want to modify the health check logic separately from strategies.

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 3]
```diff
@@ -86,17 +100,76 @@ abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
...
+                debtPayment,
+                debtOutstanding,
+                totalDebt,
+                gasCost
+            );
+        }
+
+        if(checkResult == 2) { // call fallback if loss to big
+            healthCheckFallback();
+        } else if(checkResult == 1) { // call secondary fallback if loss is acceptable
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/HealthChecker.sol:127]: This 2 and 1 checks very easy to miss. Use constants instead of it
```suggestion
        if(checkResult == SHUTDOWN_STRATEGY) { // call fallback if loss to big
            healthCheckFallback();
        } else if(checkResult == STOP_TRANSACTION) { // call secondary fallback if loss is acceptable
```

File: packages/contracts/deploy/030_USDT_ApeLendingStrategy.deploy.ts [hunk 4]
```diff
@@ -30,6 +30,7 @@ const func = deployOrUpgrade({
     chainlink__USDT_USD_feed, // asset token price feed
     6 * HOUR, // min report interval in seconds
     true, // Job is prepaid
+    1_500, // 15% threshold to shut down strategy
```
Comment (LeoVS09) [inline packages/contracts/deploy/030_USDT_ApeLendingStrategy.deploy.ts:33]: I think it must be part of new health check contracts. There need only pass address of this contract

File: packages/contracts/test/unit/BaseStrategy.t.sol [hunk 5]
```diff
@@ -450,4 +451,105 @@ contract BaseStrategyTest is TestWithERC1820Registry {
 
         assertEq(baseStrategy.minimumBetweenExecutions(), minReportInterval);
     }
+
+    function testHealthCheckPass() public {
+        uint256 _120_USDT = 120 * 10 ** 18;
```
Comment (LeoVS09) [inline packages/contracts/test/unit/BaseStrategy.t.sol:456]: Try to use fuzz testing where is possible
```suggestion
    function testHealthCheckPass(uint256 amount) public {
```

Comment (LeoVS09) [reply packages/contracts/test/unit/BaseStrategy.t.sol:456]: Same for rest of the tests

File: packages/contracts/test/unit/HealthChecker.t.sol [hunk 6]
```diff
@@ -56,7 +56,7 @@ contract HealthCheckerTest is Test {
         healthChecker.setHealthCheck(address(0));
         vm.expectCall(
             address(healthCheck),
-            abi.encodeCall(healthCheck.check, (address(0), 0, 0, 0, 0, 0))
+            abi.encodeCall(healthCheck.check, (address(0), 0, 0, 0, 0, 0, 0))
```
Comment (LeoVS09) [inline packages/contracts/test/unit/HealthChecker.t.sol:59]: Please add test which check that passed not zero params, use fuzz testing

**Review by LeoVS09 — COMMENTED** [review]
Please create a new separate contract for a health check. It will be easier to update in the future.

File: packages/contracts/deploy/030_USDT_ApeLendingStrategy.deploy.ts (hunk 4 above)
Comment (LeoVS09) [reply packages/contracts/deploy/030_USDT_ApeLendingStrategy.deploy.ts:33]: Also, please use 5% value for threshold for now

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 1 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol:18]: Nope, id doesn't reserve a memory slot. https://docs.soliditylang.org/en/latest/contracts.html#constants

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 2 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol:173]: I am really surprised to see this comment. The last time I showed you the code you were fine with the default health check logic that can be easily overridden with a custom health check from the method above.

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 7]
```diff
@@ -9,36 +9,46 @@ import {SafeInitializable} from "../upgradeable/SafeInitializable.sol";
...
+    event HealthCheckTriggered(uint8 result);
+
+    // represents 100%
+    uint256 public constant MAX_BPS = 10_000;
 
     IHealthCheck public healthCheck;
     bool public healthCheckEnabled;
 
+    // The ratio of the loss that will used to stop strategy.
+    uint public shutdownLossRatio;
```
Comment (solovev) [inline packages/contracts/src/healthcheck/HealthChecker.sol:24]:
```suggestion
    uint256 public shutdownLossRatio;
```

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 8]
```diff
@@ -86,17 +100,76 @@ abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
...
         }
+
+        if(profit > 0 || loss == 0) {
+            // if stategy profit is positive or zero, then it is healthy and we need to get this profit
+            result = 0;
+        } else if(loss > 0 && loss <= totalDebt * shutdownLossRatio / MAX_BPS) { //todo verify the condition
+            // if loss is positive but below critical loss threshold
+            result = 1;
+        } else {
+            // if loss is positive but still makes sense to run transaction to perform reporting.
```
Comment (solovev) [inline packages/contracts/src/healthcheck/HealthChecker.sol:168]: This comment is misleading, as i understand, this "else" branch tells that critical loss has occured and it's oppossite, it doesn't make sense to run transaction. Right?

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 9]
```diff
@@ -86,17 +100,76 @@ abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
...
+        if (address(strategy) == address(0)) {
             revert HealthCheckFailed();
         }
+
+        if(profit > 0 || loss == 0) {
+            // if stategy profit is positive or zero, then it is healthy and we need to get this profit
+            result = 0;
+        } else if(loss > 0 && loss <= totalDebt * shutdownLossRatio / MAX_BPS) { //todo verify the condition
+            // if loss is positive but below critical loss threshold
+            result = 1;
```
Comment (solovev) [inline packages/contracts/src/healthcheck/HealthChecker.sol:166]: Do you think we need this "intermediate" state? "Loss" - is acceptable event of the strategy behaviour. I personally ok with this, but mind the contract size, our contracts quickly become bloated with unnecessary code.

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 2 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: moved to external contract

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 3 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: added constants

File: packages/contracts/deploy/030_USDT_ApeLendingStrategy.deploy.ts (hunk 4 above)
Comment (wer32dr) [reply packages/contracts/deploy/030_USDT_ApeLendingStrategy.deploy.ts]: changed to 500 for deployment

File: packages/contracts/test/unit/BaseStrategy.t.sol (hunk 5 above)
Comment (wer32dr) [reply packages/contracts/test/unit/BaseStrategy.t.sol]: fixed.

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 7 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: fixed

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 8 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: as we agreed with Vlad even though we have the loss this loss is below the threshold and thus transaction still have to be completed. @LeoVS09 I think we need your comment here.

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 9 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: Contract size is a complex topic. Do we want to implement other technics to mitigate this issue? You mentioned diamonds or other solutions another day. I think this small portion of if else won't solve the situation because it is too small.

File: packages/contracts/src/strategies/BaseStrategy.sol [hunk 10]
```diff
@@ -158,13 +158,15 @@ abstract contract BaseStrategy is
         _adjustPosition(outstandingDebt);
 
         uint256 totalDebt = lender.currentDebt();
+        uint256 gasPrice = _gasPriceUSD() * estimatedWorkGas;
```
Comment (LeoVS09) [inline packages/contracts/src/strategies/BaseStrategy.sol:161]: Can you please make this as separate function? This logic was used a few times in separate places. And maybe more correct to name it as gasCost. 
This is minor, just suggestion

File: packages/contracts/test/unit/HealthChecker.t.sol (hunk 6 above)
Comment (LeoVS09) [reply packages/contracts/test/unit/HealthChecker.t.sol:59]: Are you not forgot about this?

File: packages/contracts/src/healthcheck/LossRatioHealthCheck.sol [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+contract LossRatioHealthCheck is SafeUUPSUpgradeable, IHealthCheck {
+    event ShutdownLossRatioChanged(uint256 ratio);
+    event HealthCheckEnabledChanged(bool enabled);
+
+    // represents 100%
+    uint256 public constant MAX_BPS = 10_000;
+
+    uint8 public constant PASS = 0;
+    uint8 public constant ACCEPTABLE_LOSS = 1;
+    uint8 public constant SIGNIFICANT_LOSS = 2;
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/LossRatioHealthCheck.sol:22]: Can you move this constants to `IHealthCheck.sol`? I think it possible to define constants on file level

File: packages/contracts/src/healthcheck/LossRatioHealthCheck.sol [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    // ------------------------------------------ Constructors ------------------------------------------
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(bool needDisableInitializers) SafeInitializable(needDisableInitializers) {} // solhint-disable-line no-empty-blocks
+
+    function initialize(uint256 _shutdownLossRatio)
+        public
+        initializer
+    {
+        __Ownable_init();
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/LossRatioHealthCheck.sol:43]: You need init `SafeUUPSUpgradeable` there, ownable will be inited under the hood.
```suggestion
        __SafeUUPSUpgradeable_init();
```

File: packages/contracts/src/healthcheck/LossRatioHealthCheck.sol [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+            // if stategy profit is positive or zero, then it is healthy and we need to get this profit
+            result = PASS;
+        } else if(loss > 0 && loss <= totalDebt * shutdownLossRatio / MAX_BPS) {
+            // if loss is positive but below critical loss threshold
+            result = ACCEPTABLE_LOSS;
+        } else {
+            // if loss is positive but still makes sense to run transaction to perform reporting.
+            result = SIGNIFICANT_LOSS;
+        }
+        return result;
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/LossRatioHealthCheck.sol:94]: I think need fix this comment, contract not know what will be done in case of significant loss. Plus maybe better use early returning pattern in this case, it will be more clear what we define as default behaivor
```suggestion
    ) external view returns (uint8)
    {
        // If no target is provided skip the execution.
        if (address(strategy) == address(0)) {
            revert HealthCheckFailed();
        }

        if(profit > 0 || loss == 0) {
            // if strategy profit is positive or zero, then it is healthy, and we need to get this profit
            return PASS;
        } 
        if(loss > 0 && loss <= totalDebt * shutdownLossRatio / MAX_BPS) {
            // if the loss is positive but below the critical loss threshold
            return ACCEPTABLE_LOSS;
        } 
        
        // If the loss is positive and bigger than we can accept mark it as significant, each strategy can handle its own way.
        return  SIGNIFICANT_LOSS;
```

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 14]
```diff
@@ -86,17 +95,35 @@ abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
...
+        if(checkResult == SIGNIFICANT_LOSS) { // call fallback if loss to big
+            healthCheckFallback();
+        } else if(checkResult == ACCEPTABLE_LOSS) { // call secondary fallback if loss is acceptable
+            acceptableLossFallback();
         }
+        // just allow to run normally in case of transaction is profitable
+    }
+
+    /// @dev Fallback function that is called when the health check fails.
+    function healthCheckFallback() internal virtual {
```
Comment (LeoVS09) [inline packages/contracts/src/healthcheck/HealthChecker.sol:119]: Little bit unclear name
```suggestion
    function healthCheckFailedFallback() internal virtual {
```

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 8 above)
Comment (LeoVS09) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: Logic is correct, just comment comment little bit unclear, I added example how to change it in my review

File: packages/contracts/src/healthcheck/HealthChecker.sol (hunk 9 above)
Comment (LeoVS09) [reply packages/contracts/src/healthcheck/HealthChecker.sol]: I think we correctly moved this to another contract. One additional if for the intermediate state, is not a big deal, but will save as much when we will want to enhance this logic. We can remove logic which handle acceptable loss on strategy side, it will decrease sise on few lines but it not a big deal. My personal preference keep it as it is for now

File: packages/contracts/test/unit/HealthChecker.t.sol (hunk 6 above)
Comment (wer32dr) [reply packages/contracts/test/unit/HealthChecker.t.sol:59]: Nope, that change just doesn't make sense since that tests the empty implementation of IHealthCheck rather than normal HealtchCheck implementation. Could you elaborate on why I should reformat the code for the testing mock?

File: packages/contracts/src/healthcheck/HealthChecker.sol [hunk 15]
```diff
@@ -9,8 +9,17 @@ import {SafeInitializable} from "../upgradeable/SafeInitializable.sol";
 error HealthCheckFailed();
 
 abstract contract HealthChecker is SafeInitializable, OwnableUpgradeable {
-    event HealthCheckChanged(address);
-    event HealthCheckEnabledChanged(bool);
+    event HealthCheckChanged(address healthCheck);
+    event ShutdownLossRatioChanged(uint256 ratio);
```
[the remark this answers is not in the document]
Comment (wer32dr) [reply packages/contracts/src/healthcheck/HealthChecker.sol:13]: Both? I could agree with ShutdownLossRatioChanged but I'd prefer to keep the rest.

File: packages/contracts/src/healthcheck/LossRatioHealthCheck.sol (hunk 11 above)
Comment (wer32dr) [reply packages/contracts/src/healthcheck/LossRatioHealthCheck.sol:22]: That is the interface so it doesn't support constants. I can convert it to an abstract class but it will be a bit weird. Another option is to have a separate contract with constants only.

Comment (wer32dr) [reply packages/contracts/src/healthcheck/LossRatioHealthCheck.sol:22]: Nope, I was wrong it works on file level.

File: packages/contracts/test/unit/HealthChecker.t.sol (hunk 6 above)
Comment (LeoVS09) [reply packages/contracts/test/unit/HealthChecker.t.sol]: Yes, if you test case `abi.encodeCall(healthCheck.check, (address(0), 1, 2, 3, 4, 5, 6))` and check that all params correctly passed, then we will be sure that there no misplaced variables. They all unit256 and all in one line, it easy to pass one param in incorrect place

Comment (LeoVS09) [reply packages/contracts/test/unit/HealthChecker.t.sol]: Yes, if you test case abi.encodeCall(healthCheck.check, (address(0), 1, 2, 3, 4, 5, 6)) and check that all params correctly passed, then we will be sure that there no misplaced variables. They all unit256 and all in one line, it easy to pass one param in incorrect place

**Review by LeoVS09 — APPROVED** [review]
